### PR TITLE
Restore covariant comparison to bluebird's Promise

### PR DESCRIPTION
**Pull request by sandersn**
Now that typescript@next checks variance more strictly for conditional types, bluebird's Promise is invariant. This PR removes conditional types from bluebird so that Promise compares covariantly again.

This is the correct follow-up to my initial cleanup attempt. Basically, we had to remove any conditional types from Bluebird since conditional types are checked more strictly now. They are replaced with (1) overloads (2) this-parameters (3) inference to intersections.

The combination is harder to read than conditional types. And it required some duplicative changes to bluebird-global, but its source package hasn't been updated in 2 years so I think that is fine.

### CODE REVIEW

File: types/knex/knex-tests.ts [hunk 1]
```diff
@@ -1111,6 +1111,7 @@ knex('users')
 //
 // Migrations
 //
+const name = "test";
```
Comment (weswigham) [inline types/knex/knex-tests.ts]: ???

Comment (sandersn) [reply types/knex/knex-tests.ts]: The test otherwise to refers to the global `name`, which the linter doesn't like.

**Comment by sandersn** [comment]
@lhecker This is the correct follow-up to my initial cleanup attempt. Basically, we had to remove any conditional types from Bluebird since conditional types are checked more strictly now. They are replaced with (1) overloads (2) this-parameters (3) inference to intersections. 

The combination is harder to read than conditional types. And it required some changes to bluebird-global, but the source package hasn't been updated in 2 years so I think that is fine.

File: types/bluebird/index.d.ts [hunk 2]
```diff
@@ -578,42 +581,59 @@ declare class Bluebird<R> implements PromiseLike<R>, Bluebird.Inspection<R> {
...
+  /**
+   * Same as calling `Promise.any(thisPromise)`. With the exception that if this promise is bound to a value, the returned promise is bound to that value too.
+   */
+  any(): Bluebird<never>;
+
+  /**
+   * Same as calling `Promise.some(thisPromise)`. With the exception that if this promise is bound to a value, the returned promise is bound to that value too.
+   * Same as calling `Promise.some(thisPromise)`. With the exception that if this promise is bound to a value, the returned promise is bound to that value too.
+   */
+  some(this: Bluebird<Iterable<{}>>, count: number): Bluebird<R>;
```
Comment (weswigham) [inline types/bluebird/index.d.ts]: shouldn't this be `some<Q>(this: Bluebird<R & Iterable<Q>>, count: number): Bluebird<Q>;` ? Or do I misunderstand?

Comment (sandersn) [reply types/bluebird/index.d.ts]: Not according to the tests:

```ts
let fooArrProm: Promise<Foo[]>;
fooArrProm = fooArrProm.some(num);
```

It doesn't make sense to me either.

Comment (weswigham) [reply types/bluebird/index.d.ts]: Update: I do misunderstand, and the `count` parameter is important.

File: types/bluebird-global/index.d.ts [hunk 3]
```diff
@@ -139,17 +142,17 @@ declare global {
         isRejected: Bluebird<T>["isRejected"];
         isResolved: Bluebird<T>["isResolved"];
         lastly: Bluebird<T>["lastly"];
-        map: Bluebird<T>["map"];
-        mapSeries: Bluebird<T>["mapSeries"];
+        map<U, Q>(this: Promise<T & Iterable<Q>>, mapper: IterateFunction<Q, U>, options?: Bluebird.ConcurrencyOption): Bluebird<U[]>;
+        mapSeries<U, Q>(this: Promise<T & Iterable<Q>>, iterator: IterateFunction<Q, U>): Bluebird<U[]>;
         nodeify: Bluebird<T>["nodeify"];
         props: Bluebird<T>["props"];
         race: Bluebird<T>["race"];
```
Comment (lhecker) [inline types/bluebird-global/index.d.ts]: I believe you need to change `race` as well?
The bluebird definition is:
```ts
race<Q>(this: Bluebird<R & Iterable<Q>>): Bluebird<Q>;
```

Comment (sandersn) [reply types/bluebird-global/index.d.ts]: Fixed, nice catch.

**Pull request merged**